In [ ]:
python -c "import sys; print(sys.executable)"

(.venv) PS C:\dev\llm-data-analysis-study> python -c "import sys; print(sys.executable)"
C:\dev\llm-data-analysis-study\.venv\scripts\python.exe
(.venv) PS C:\dev\llm-data-analysis-study> 

In [4]:
from pathlib import Path
import sys

print(sys.executable)
print(Path.cwd())

c:\Users\korea\AppData\Local\Python\pythoncore-3.14-64\python.exe
c:\dev\llm-data-analysis-study\chapter03


In [6]:
from pathlib import Path


def find_project_root(start: Path) -> Path:
    start = start.resolve()

    for candidate in [start, *start.parents]:
        if (
            (candidate / "data" / "raw").exists()
            and (candidate / "book").exists()
        ):
            return candidate

    raise FileNotFoundError(
        "프로젝트 루트를 찾지 못했습니다. "
        "Notebook을 저장소 안에서 실행해 주세요."
    )


PROJECT_ROOT = find_project_root(Path.cwd())
DATA_DIR = PROJECT_ROOT / "data" / "raw"

print(PROJECT_ROOT)
print(DATA_DIR)

C:\dev\llm-data-analysis-study
C:\dev\llm-data-analysis-study\data\raw


In [7]:
expected_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",
]

file_check = {
    filename: (DATA_DIR / filename).exists()
    for filename in expected_files
}

file_check

{'customers.csv': True,
 'products.csv': True,
 'orders.csv': True,
 'order_items.csv': True}

In [41]:
import pandas as pd

customers = pd.read_csv(DATA_DIR / "customers.csv")
products = pd.read_csv(DATA_DIR / "products.csv")
orders = pd.read_csv(DATA_DIR / "orders.csv")
order_items = pd.read_csv(DATA_DIR / "order_items.csv")

datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

In [9]:
for name, df in datasets.items():
    print(name, df.shape)

customers (150, 6)
products (100, 4)
orders (300, 5)
order_items (764, 5)


In [12]:
customers.head()

,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-06-19
1,2,김정호,F,32,대구,2025-11-02
2,3,이경수,F,61,성남,2024-06-12
3,4,조영호,F,55,울산,2026-04-13
4,5,이예원,F,19,부산,2024-09-13


In [13]:
customers.tail()

,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2025-12-23
146,147,이정남,M,19,부산,2025-02-11
147,148,오도현,M,29,고양,2026-06-15
148,149,김정자,M,20,부산,2024-10-19
149,150,조미영,M,40,대전,2025-12-04


In [14]:
products.head()

,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


In [15]:
orders.head()

,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-05-07,card,completed
1,2,77,2025-07-23,naver_pay,cancelled
2,3,138,2025-11-19,bank_transfer,cancelled
3,4,57,2026-01-30,kakao_pay,cancelled
4,5,125,2025-12-21,card,cancelled


In [16]:
order_items.head()

,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


In [17]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    print(list(df.columns))
    print(df.dtypes)


===== customers =====
['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object

===== products =====
['product_id', 'product_name', 'category', 'price']
product_id      int64
product_name      str
category          str
price           int64
dtype: object

===== orders =====
['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
order_id          int64
customer_id       int64
order_date          str
payment_method      str
order_status        str
dtype: object

===== order_items =====
['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']
order_item_id    int64
order_id         int64
product_id       int64
quantity         int64
unit_price       int64
dtype: object


In [18]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB


In [19]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    print(df.isna().sum())


[customers]
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

[products]
product_id      0
product_name    0
category        0
price           0
dtype: int64

[orders]
order_id          0
customer_id       0
order_date        0
payment_method    0
order_status      0
dtype: int64

[order_items]
order_item_id    0
order_id         0
product_id       0
quantity         0
unit_price       0
dtype: int64


In [20]:
(customers.isna().mean() * 100).round(2)

customer_id    0.0
name           0.0
gender         0.0
age            0.0
city           0.0
signup_date    0.0
dtype: float64

In [21]:
for name, df in datasets.items():
    print(name, df.duplicated().sum())

customers 0
products 0
orders 0
order_items 0


In [22]:
key_columns = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}

for name, key in key_columns.items():
    df = datasets[name]
    print(
        name,
        "missing:", df[key].isna().sum(),
        "duplicated:", df[key].duplicated().sum(),
    )

customers missing: 0 duplicated: 0
products missing: 0 duplicated: 0
orders missing: 0 duplicated: 0
order_items missing: 0 duplicated: 0


In [25]:
customers[["age"]].describe()

,age
count,150.000000
mean,42.086667
std,15.613166
min,19.000000
25%,29.000000
50%,40.000000
75%,57.000000
max,69.000000


In [26]:
products[["price"]].describe()

,price
count,100.000000
mean,110040.000000
std,56433.910574
min,5000.000000
25%,65750.000000
50%,112000.000000
75%,161000.000000
max,200000.000000


In [27]:
order_items[["quantity", "unit_price"]].describe()

,quantity,unit_price
count,764.000000,764.000000
mean,3.053665,108561.518325
std,1.410873,56996.770604
min,1.000000,5000.000000
25%,2.000000,62000.000000
50%,3.000000,111000.000000
75%,4.000000,161250.000000
max,5.000000,200000.000000


In [28]:
customers["city"].value_counts(dropna=False)

city
성남    21
광주    17
부산    16
대구    15
서울    15
울산    14
인천    14
대전    14
수원    13
고양    11
Name: count, dtype: int64

In [29]:
products["category"].value_counts(dropna=False)

category
스포츠     19
전자기기    17
생활용품    16
뷰티      16
도서      14
패션      11
식품       7
Name: count, dtype: int64

In [30]:
orders["order_status"].value_counts(dropna=False)

order_status
completed    184
cancelled     64
refunded      52
Name: count, dtype: int64

In [31]:
orders["order_date"] = pd.to_datetime(
    orders["order_date"],
    errors="coerce",
)

print("변환 실패:", orders["order_date"].isna().sum())
print("가장 빠른 주문일:", orders["order_date"].min())
print("가장 최근 주문일:", orders["order_date"].max())

변환 실패: 0
가장 빠른 주문일: 2025-07-09 00:00:00
가장 최근 주문일: 2026-07-08 00:00:00


In [32]:
signup_check = pd.to_datetime(
    customers["signup_date"],
    errors="coerce",
)

print("가입일 변환 실패:", signup_check.isna().sum())

가입일 변환 실패: 0


In [33]:
invalid_customers = orders.loc[
    ~orders["customer_id"].isin(customers["customer_id"])
]

invalid_orders = order_items.loc[
    ~order_items["order_id"].isin(orders["order_id"])
]

invalid_products = order_items.loc[
    ~order_items["product_id"].isin(products["product_id"])
]

print("없는 customer_id:", len(invalid_customers))
print("없는 order_id:", len(invalid_orders))
print("없는 product_id:", len(invalid_products))

없는 customer_id: 0
없는 order_id: 0
없는 product_id: 0


In [38]:
def check_data_overview(name, df, key_column=None):
    print(f"===== {name} =====")
    print("shape:", df.shape)
    print("columns:", list(df.columns))
    print("dtypes:")
    print(df.dtypes)
    print("missing:")
    print(df.isna().sum())
    print("duplicated rows:", df.duplicated().sum())

    if key_column is not None:
        print("key missing:", df[key_column].isna().sum())
        print("key duplicated:", df[key_column].duplicated().sum())


check_data_overview("주문",orders)

===== 주문 =====
shape: (300, 5)
columns: ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
dtypes:
order_id                   int64
customer_id                int64
order_date        datetime64[us]
payment_method               str
order_status                 str
dtype: object
missing:
order_id          0
customer_id       0
order_date        0
payment_method    0
order_status      0
dtype: int64
duplicated rows: 0


customers.csv 파일의 구조 요약 및 분석 전 추가 확인 사항입니다.

📊 customers.csv 구조 요약
행/열: 150행, 6열 (150, 6)

컬럼: customer_id, name, gender, age, city, signup_date

데이터 타입:

customer_id: int64 (정수형)

name: object (문자열)

gender: object (문자열)

age: int64 (정수형)

city: object (문자열)

signup_date: object (문자열, 날짜 형태)

결측치: 없음 (모든 컬럼 0건)

전체 중복: 없음 (완전 일치하는 행 0건)

customer_id 중복: 없음 (고유 식별자 중복 0건)

🔍 분석 전 추가 확인 사항 (구분)
데이터 분석을 본격적으로 시작하기 앞서, 단순한 추정(가정)과 데이터에서 실제 확인이 필요한 항목을 구분하여 점검해야 합니다.

1. 실제 확인이 필요한 항목 (데이터 검증)
날짜 타입 변환 여부 (signup_date): 현재 object 타입으로 되어 있으므로, 시계열 분석이나 기간별 필터링을 위해 datetime 타입으로 변환이 필요한지 확인해야 합니다.

값의 유효성 범위 점검:

age(나이): 최소 19세에서 최대 69세로 확인되나, 비정상적인 값(음수나 터무니없는 고령 등)이 숨어있지 않은지 분포를 상세히 확인해야 합니다.

gender(성별): 데이터상 ['F', 'M'] 값만 존재하지만, 공백이나 대소문자 혼용 여부를 명확히 체크해야 합니다.

city(도시): 총 10개의 도시가 포함되어 있으며, 표기 방식(띄어쓰기, 오탈자 등)에 일관성이 있는지 확인해야 합니다.

2. 추정 및 비즈니스 맥락 확인 항목 (추가 정보 필요)
데이터의 수집 기준 및 범위: 이 150명의 고객 데이터가 전체 고객을 대표하는 표본인지, 특정 기간이나 특정 이벤트로 유입된 코호트인지 추정하지 말고 정의를 명확히 해야 합니다.

외래키(FK) 연관성: 앞서 다룬 orders나 order_items 테이블과 조인(Join)할 때, customer_id가 누락 없이 완벽하게 매핑되는지(orders의 customer_id가 본 파일의 ID 범주 내에 완전히 속하는지) 실제 교차 검증이 필요합니다.

In [47]:
def check_data(name, df):
    print(f"{name} 구조 요약")
    print(f"행/열: {df.shape[0]} / {df.shape[1]}")
    print("컬럼:", list(df.columns))
    print("데이터 타입:")
    print(df.dtypes)
    print("결측치:")
    print(df.isna().sum())
    print("customer_id 중복:", df.duplicated(subset=["customer_id"]).sum())

check_data("customers.csv", customers)

customers.csv 구조 요약
행/열: 150 / 6
컬럼: ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
데이터 타입:
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object
결측치:
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64
customer_id 중복: 0


[Chapter 03 Evidence]

1. CSV 파일 존재 여부
- customers.csv: PASS
- products.csv: PASS
- orders.csv: PASS
- order_items.csv: PASS

2. 데이터 규모
- customers: 150 rows × 6 cols
- products: 100 rows × 4 cols
- orders: 300 rows × 5 cols
- order_items: 764 rows × 5 cols

3. 품질 점검
- 주요 ID 결측: 0
- 주요 ID 중복: 0
- 전체 행 중복: 0
- 날짜 변환 실패: 0

4. 관계 검증
- 없는 customer_id: 0
- 없는 order_id: 0
- 없는 product_id: 0

5. LLM 활용
- 요청 목적: Data file의 구조를 확인하기 위해
- 실제 반영: 분석 전 추가 확인 사항 (구분)
데이터 분석을 본격적으로 시작하기 앞서, 단순한 추정(가정)과 데이터에서 실제 확인이 필요한 항목을 구분하여 점검해야 합니다.
- 사람이 검증한 항목: 데이터 분석에 용이한 형태로 자료가 작성되었는 지에 대한 검증